### PR TITLE
add test coverage CI and badge

### LINKED ISSUE #42: test coverage
This crate currently has zero test coverage

This makes refactoring and continuous integration impossible.
It also makes it extremely difficult to support community contributions

Perhaps an incremental approach is best? Start tracking test coverage, then insist that each pull request going forward either improve test coverage, or at least not significantly degrade it.

### PR DESCRIPTION
**Pull request by danieleades**
closes #42

This pull request is not ready to be merged, as it requires a codecov account to be created, and the api token to be added to the repo as a 'github secret'

### CODE REVIEW

**Comment by danieleades** [comment]
@Hirevo this is good to go, just needs a codecov token added as a 'secret'. Is this something that you want?

File: .github/workflows/coverage.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+jobs:
+  coverage:
+    name: Test Coverage
+    runs-on: ubuntu-latest
+    steps:
+
+      - name: checkout source
+        uses: actions/checkout@v2
+
+      - name: Install nightly toolchain
```
[the remark this answers is not in the document]
Comment (danieleades) [reply .github/workflows/coverage.yml:14]: That's a good question, I'm not actually sure. As far as I know tarpaulin doesn't require nightly. This is a copy and paste from another project of mine (which *does* rely on nightly features).
I'll switch this to stable and see if there's any difference

**Comment by danieleades** [comment]
You might have to play around with the codecov settings, but if you like, you can have codecov comment on pull requests with a breakdown of how the pull request will affect the test coverage.
It's a handy way to ratchet up coverage over time

**Comment by danieleades** [comment]
the codecov action is failing to upload to codecov. not sure why

**Comment by danieleades** [comment]
that's got it! should be good to merge now

**Comment by Hirevo** [comment]
Looks good to me as well !  
Many thanks for the help !

**Pull request merged**